### ISSUE TITLE
scripts that omit a shebang can fail to exec if given > a number of arguments

### ISSUE THREAD

**Issue by floam** [fish-shell/fish-shell#8437]
```fish
$ echo 'echo $@ | wc -w' > ~/count.sh && chmod +x ./count.sh
$ ~/count.sh *
      88
$ cd $TMPDIR
$ ~/count.sh
exec: The file './test.sh' is marked as an executable but could not be run by the operating system.
$ ls | wc -l
     900
```

Works in csh, bash, zsh, ksh.

**Comment by floam** [comment]
https://github.com/fish-shell/fish-shell/blob/master/src/exec.cpp#L164

**Comment by faho** [comment]
This should probably be using ARG_MAX. Unfortunately it seems actually trying `sysconf(_SC_ARG_MAX)` doesn't work in this context (does that allocate or something?), so we'd have to do it before and then just read it.

Can that value change over the lifetime of a process? Is that something we'd have to account for?

**Comment by floam** [comment]
Are you seeing anything on the heap for `sysconf`?

**Comment by faho** [comment]
I have no idea how to check such a thing.

**Comment by floam** [comment]
I'm not sure if _SC_ARG_MAX itself can change, but the actual usable maximum argument length definitely changes over the lifetime of a process. On linux, the effective usable size is that number minus the length of `environ`/`envp`. POSIX says to take off 2048 for good measure.

**Comment by faho** [comment]
Yes, but there is no need to be that exact. We can simply allocate ARG_MAX and if the env puts us over the usable value we'll get an error later. That's okay.

**Comment by floam** [comment]
Should not allocate here.

**Comment by floam** [comment]
I think

**Comment by floam** [comment]
I notice other shells do not bother to calculate it as @faho is suggesting. I wonder if we can dispense with doing it in the spot where we do. Utilities like xargs do.

**Comment by faho** [comment]
>Should not allocate here.

"allocate" in the sense of "have an array of that size that we pass to execve".

**Comment by floam** [comment]
Nevermind - that calculation is just for the sake of error reporting.
